### PR TITLE
NXP-29984: Runtime Registries Migrations #3

### PR DESCRIPTION
**Pull request by atchertchian**

### CODE REVIEW

**Comment by atchertchian** [comment]
With review ElasticSearchComponent migration as rest-api tests show that embedded server config can be null. Will consider client config is mandatory, though (will throw a meaningful exception if so)

File: modules/core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/blob/DocumentBlobManagerComponent.java [hunk 1]
```diff
@@ -295,7 +265,7 @@ public void updateBlob(Document doc, Consumer<BlobUpdateContext> contextFiller)
         }
         BlobProvider blobProvider = Framework.getService(BlobManager.class).getBlobProvider(blob);
         if (blobProvider == null) {
-            log.error("No blob provider found for blob: " + blob.getKey());
+            log.error("No blob provider found for blob: {}", blob.getKey());
```
Comment (efge) [inline modules/core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/blob/DocumentBlobManagerComponent.java:268]: `blob::getKey`
(the perf aspect is trivial, but you may get warning from some code quality tools if you don't use constants (or a method handle or a lambda) in log statements)

File: modules/core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/adapter/DocumentAdapterService.java [hunk 2]
```diff
@@ -40,62 +39,37 @@
...
-    }
-
     @Override
-    public void activate(ComponentContext context) {
+    public void start(ComponentContext context) {
         adapters = new ConcurrentHashMap<>();
+        adapters.putAll(
+                this.<DocumentAdapterDescriptor> getRegistryContributions(XP)
+                    .stream()
+                    .collect(Collectors.toConcurrentMap(DocumentAdapterDescriptor::getInterface, Function.identity())));
```
Comment (efge) [inline modules/core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/adapter/DocumentAdapterService.java:55]: Why not just assign directly, no need for `putAll`. Also if nobody changes the map after creation the concurrent aspect isn't needed.

File: modules/platform/nuxeo-platform-filemanager/src/main/java/org/nuxeo/ecm/platform/filemanager/service/FileManagerRegistry.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    public void unregister(String tag) {
+        registries.values().stream().filter(r -> r.isTagged(tag)).forEach(r -> r.unregister(tag));
+    }
+
+    public <T> List<T> getContributionValues(Class<?> klass) {
+        String rid = computePluginsExtensionPoint(klass);
+        if (!registries.containsKey(rid)) {
+            return Collections.emptyList();
+        }
+        return registries.get(rid).getContributionValues();
```
Comment (efge) [inline modules/platform/nuxeo-platform-filemanager/src/main/java/org/nuxeo/ecm/platform/filemanager/service/FileManagerRegistry.java:83]: `Map.containsKey()` followed by `Map.get()` is not optimal (unless the map can contain null values).

          MapRegistry registry = registries.get(rid);
          return registry == null ? List.of() : registry.getContributionValues();

File: modules/platform/nuxeo-platform-filemanager/src/main/java/org/nuxeo/ecm/platform/filemanager/service/FileManagerService.java [hunk 4]
```diff
@@ -121,82 +120,78 @@
...
-        String xp = computePluginsExtensionPoint(CreationContainerListProviderDescriptor.class);
-        creationContainerListProviders = getDescriptors(xp).stream()
-                                                           .map(CreationContainerListProviderDescriptor.class::cast)
-                                                           .map(CreationContainerListProviderDescriptor::newInstance)
-                                                           .collect(Collectors.toList());
+    protected void registerPlugins() {
+        FileManagerRegistry registry = getExtensionPointRegistry(PLUGINS_EP);
+        fileImporters = registry.getContributionValues(FileImporterDescriptor.class)
+                                .stream()
+                                .map(FileImporterDescriptor.class::cast)
```
Comment (efge) [inline modules/platform/nuxeo-platform-filemanager/src/main/java/org/nuxeo/ecm/platform/filemanager/service/FileManagerService.java:164]: Is the cast needed? `getContributionValues` should returned objects with the proper type already.

File: modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/exceptionhandling/service/ExceptionHandlingComponent.java [hunk 5]
```diff
@@ -15,68 +15,65 @@
...
+            parameters.setHandlers(desc.getMessages());
+            parameters.setLoggerName(desc.getLoggerName());
+            parameters.setDefaultErrorPage(desc.getDefaultPage());
+        });
+        this.<RequestDumpDescriptor> getRegistryContribution(ExtensionPoint.requestdump.name()).ifPresent(desc -> {
+            RequestDumper dumper = newInstance(desc.getKlass());
+            dumper.setNotListedAttributes(desc.getAttributes());
+            parameters.setRequestDumper(dumper);
+        });
+        this.<ListenerDescriptor> getRegistryContribution(ExtensionPoint.listener.name()).ifPresent(ld -> {
```
Comment (efge) [inline modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/exceptionhandling/service/ExceptionHandlingComponent.java:68]: `ld` -> `desc` for consistency

File: modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/ElasticSearchComponent.java [hunk 6]
```diff
@@ -106,74 +106,40 @@
...
+        ElasticSearchEmbeddedServerConfig embeddedServerConfig = this.<ElasticSearchEmbeddedServerConfig> getRegistryContribution(
+                EP_EMBEDDED_SERVER).orElse(null);
+        ElasticSearchClientConfig clientConfig = this.<ElasticSearchClientConfig> getRegistryContribution(
+                EP_CLIENT_INIT)
+                                                     .orElseThrow(() -> new IllegalArgumentException(
+                                                             "Missing Elasticsearch client contribution"));
+        Map<String, ElasticSearchIndexConfig> indexConfig = this.<MapRegistry> getExtensionPointRegistry(EP_INDEX)
+                                                                .getContributions();
+        JsonESDocumentWriter jsonESDocumentWriter = this.<ElasticSearchDocWriterDescriptor> getRegistryContribution(
+                EP_DOC_WRITER).map(writerDescriptor -> {
```
Comment (efge) [inline modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/ElasticSearchComponent.java:135]: `writerDescriptor` -> `desc` for consistency maybe?

File: modules/platform/nuxeo-dmk-adaptor/src/main/java/org/nuxeo/dmk/DmkComponent.java [hunk 7]
```diff
@@ -116,19 +115,20 @@ protected void destroyConnector(JDMKServerConnector connector) {
...
         if (configs.containsKey("html")) {
             htmlAdaptor = newAdaptor(configs.get("html"));
             log.info("JMX HTML adaptor available at port 8081 (not active, to be started in JMX console)");
         }
         if (configs.containsKey("http")) {
             httpConnector = newConnector(configs.get("http"));
-            log.info("JMX HTTP connector available at " + httpConnector.getAddress()
-                    + " (not active, to be started in JMX console)");
+            log.info("JMX HTTP connector available at {} (not active, to be started in JMX console)",
+                    httpConnector.getAddress());
```
Comment (efge) [inline modules/platform/nuxeo-dmk-adaptor/src/main/java/org/nuxeo/dmk/DmkComponent.java:126]: `httpConnector::getAddress`

File: modules/platform/nuxeo-platform-document-routing/nuxeo-routing-core/src/main/java/org/nuxeo/ecm/platform/routing/core/impl/DocumentRoutingServiceImpl.java [hunk 8]
```diff
@@ -143,51 +139,58 @@
...
             } catch (ReflectiveOperationException e) {
                 throw new NuxeoException(e);
             }
-        } else if (ROUTE_MODELS_IMPORTER_XP.equals(extensionPoint)) {
-            RouteModelResourceType res = (RouteModelResourceType) contribution;
-            registerRouteResource(res, contributor.getRuntimeContext());
-        }
+        });
+        routeTemplates = getRegistryContributions(ROUTE_MODELS_IMPORTER_XP).stream()
+                                                                           .map(RouteModelResourceType.class::cast)
```
Comment (efge) [inline modules/platform/nuxeo-platform-document-routing/nuxeo-routing-core/src/main/java/org/nuxeo/ecm/platform/routing/core/impl/DocumentRoutingServiceImpl.java:169]: Could avoid cast by doing `this.<RouteModelResourceType> getRegistryContributions` ?

**Review by efge — CHANGES_REQUESTED** [review]

File: modules/platform/nuxeo-platform-filemanager/src/main/java/org/nuxeo/ecm/platform/filemanager/service/FileManagerService.java (hunk 4 above)
Comment (atchertchian) [reply modules/platform/nuxeo-platform-filemanager/src/main/java/org/nuxeo/ecm/platform/filemanager/service/FileManagerService.java:164]: It is needed, i cannot find any way of using `this.<FileImporter>` in this case.
A working alternative would be:
```
List<FileImporterDescriptor> fdescs = registry.getContributionValues(FileImporterDescriptor.class);
fileImporters = fdescs.stream()
                      .map(FileImporterDescriptor::newInstance)
                      .collect(Collectors.toMap(FileImporter::getName, Function.identity()));
```
Do you prefer this way or have a better alternative?

Comment (efge) [reply modules/platform/nuxeo-platform-filemanager/src/main/java/org/nuxeo/ecm/platform/filemanager/service/FileManagerService.java:164]: Keep what you have if it works, it's not very important.

Comment (atchertchian) [reply modules/platform/nuxeo-platform-filemanager/src/main/java/org/nuxeo/ecm/platform/filemanager/service/FileManagerService.java:164]: I'd actually be curious to understand why it does not work in this case, as the custom registry API is the same than the usual helper method (and my eclipse compile does not accept use of `this.<FileImporter>` here, I don't see the logic)

Comment (efge) [reply modules/platform/nuxeo-platform-filemanager/src/main/java/org/nuxeo/ecm/platform/filemanager/service/FileManagerService.java:164]: `this.<FileImporterDescriptor>` you mean?

Comment (atchertchian) [reply modules/platform/nuxeo-platform-filemanager/src/main/java/org/nuxeo/ecm/platform/filemanager/service/FileManagerService.java:164]: Yes

Comment (atchertchian) [reply modules/platform/nuxeo-platform-filemanager/src/main/java/org/nuxeo/ecm/platform/filemanager/service/FileManagerService.java:164]: Actually compile complains on "." of `registry.`, i guess that's why (usually helper method is called directly)

Comment (efge) [reply modules/platform/nuxeo-platform-filemanager/src/main/java/org/nuxeo/ecm/platform/filemanager/service/FileManagerService.java:164]: I'll test once you push and I can have it in my IDE

**Review by kevinleturc — APPROVED** [review]
Just a detail 👌.

File: modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/exceptionhandling/service/ExceptionHandlingComponent.java [hunk 9]
```diff
@@ -15,68 +15,65 @@
...
-            ListenerDescriptor ld = (ListenerDescriptor) contribution;
-            exceptionHandlerParameters.setListener(newInstance(ld.getKlass()));
-            break;
-        default:
-            throw new RuntimeException("error in exception handling configuration");
-        }
+    public void start(ComponentContext context) {
+        this.<ExceptionHandlerDescriptor> getRegistryContribution(ExtensionPoint.exceptionhandler.name())
+            .ifPresentOrElse(desc -> exceptionHandler = newInstance(desc.getKlass()),
+                    () -> exceptionHandler = new DefaultNuxeoExceptionHandler());
```
[the remark this answers is not in the document]
Comment (atchertchian) [reply modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/exceptionhandling/service/ExceptionHandlingComponent.java:55]: I actually tried that first, i get the following compile error:

```
The method orElse(capture#5-of ? extends NuxeoExceptionHandler) in the type Optional<capture#5-of ? extends NuxeoExceptionHandler> is not applicable for the arguments (NuxeoExceptionHandler)
```

[the remark this answers is not in the document]
Comment (kevinleturc) [reply modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/exceptionhandling/service/ExceptionHandlingComponent.java:55]: Too many generics 😄.

**Review by efge — APPROVED** [review]

**Pull request merged**